README.md
# tftp plugin

A lib for Trivial File Transfer Protocol(TFTP) server and client.

## How To Use

### Server

We can start a tftp server, and then a tftp client could download/upload file from server.

The server has `onRead`,`onWrite` and `onError` event.

For example:

```dart
  var server = await TFtpServer.bind("0.0.0.0", port: 6699);
  server.listen((socket) {
    socket.listen(onRead: (file, onProcess) {
      print("read file from server:$file");
      onProcess(progressCallback: (count, total) {
        print("read:${count / total}%");
      });
      return "$readAblePath/$file"; // server will read file from return value
    }, onWrite: (file, doTransform) {
      doTransform();
      infoController.add("write file to server:$file");
      return "$writeAblePath/$file"; // server will write file to return value
    }, onError: (code, msg) {
      infoController.add("Error[$code]:$msg");
    });
  });
```

### client

We can create a tftp client to download/upload file to a tftp server.

1. init

> var client = await TFtpClient.bind("0.0.0.0", port);

2. Get 

> client.get(localFilePath, remoteFilePath,remoteHost, remotePort);

3. Put

> client.put(localFilePath, remoteFilePath,remoteHost, remotePort);

### Topic:

If you want to read/write file to `EXTERNAL STORAGE`,you should 
Add permissions in `AndroidManifest.xml`

```xml
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
```

CHANGELOG.md
## [1.0.0] - Release tftp

* Release a simple tftp server and tftp client

## [1.0.1] - fix bug.

* fix bug.
* format code.
* fix document.

## [1.1.0] - support null safe.

* support null safe.
* add analysis.
* 
## [1.2.0] - fix fixed length list issue and refactor example.

* fix lint warnings
* fix fixed-length list issue
* refactor example app
* migrate to Android V2 embedding and AndroidX
* ignore generated files

test/tftp_server_test.dart
import 'package:flutter/foundation.dart';
import 'package:tftp/tftp.dart';

void main() {
  TFtpServer.bind("127.0.0.1", port: 6699).then((server) {
    server.listen((socket) {
      socket.listen(onRead: (file, onProcess) {
        onProcess(progressCallback: (count, total) {
          if (kDebugMode) {
            print("$count/$total");
          }
        });
        return file;
      }, onWrite: (file, doTransform) {
        doTransform(overwrite: false);
        return file;
      });
    });
  });
}


lib/src/client.dart
import 'dart:async';
import 'dart:io';

import 'package:tftp/src/common.dart';

class TFtpClient {
  final String host;
  final int port;
  final int blockSize;

  late RawDatagramSocket _socket;
  late Stream<RawSocketEvent> _stream;

  List<int> _receivedBlock = <int>[];

  TFtpClient(this.host, this.port, {this.blockSize = 512});

  static Future<TFtpClient> bind(String host, int port, {int blockSize = 512}) {
    Completer<TFtpClient> completer = Completer();
    RawDatagramSocket.bind(host, port).then((socket) {
      var client = TFtpClient(host, port, blockSize: blockSize);
      client._socket = socket;
      client._stream = socket.asBroadcastStream();
      completer.complete(client);
    });
    return completer.future;
  }

  Future put(
      String localFile, String remoteFile, String remoteAddress, int remotePort,
      {ProgressCallback? progressCallback, ErrorCallBack? onError}) async {
    List<int> sendPacket = [
      [0, OpCode.WRQ_VALUE],
      encoder.convert(remoteFile),
      [0],
      TransType.octet,
      [0],
    ].expand((x) => x).toList();
    var blockNum = 0;
    var rp = await _sendPacket(sendPacket, blockNum, remoteAddress, remotePort);

    RandomAccessFile fileWait2Write = await File(localFile).open();
    var totalSize = fileWait2Write.lengthSync();
    List<int> dataBlock;
    while ((dataBlock = await fileWait2Write.read(blockSize)).isNotEmpty) {
      ++blockNum;
      blockNum = blockNum > 65535 ? 0 : blockNum;

      List<int> sendPacket = [
        [0, OpCode.DATA_VALUE],
        [blockNum >> 8, blockNum & 0xff],
        dataBlock
      ].expand((x) => x).toList();

      await _sendPacket(sendPacket, blockNum, remoteAddress, rp);
      if (blockNum * blockSize == totalSize) {
        ++blockNum;
        blockNum = blockNum > 65535 ? 0 : blockNum;
        List<int> sendPacket = [
          [0, OpCode.DATA_VALUE],
          [blockNum >> 8, blockNum & 0xff],
        ].expand((x) => x).toList();
        await _sendPacket(sendPacket, blockNum, remoteAddress, remotePort);
      }
    }
  }

  Future get(
      String localFile, String remoteFile, String remoteAddress, int remotePort,
      {ErrorCallBack? onError}) async {
    Completer<int> completer = Completer();

    var file = File(localFile);
    var io = file.openWrite();
    late StreamSubscription subscription;
    subscription = _stream.listen((ev) {
      if (RawSocketEvent.read == ev) {
        var data = _socket.receive();
        var port = data!.port;
        if (data.data[1] == OpCode.DATA_VALUE) {
          var blockSeq = (data.data[2] << 8) + data.data[3];
          if (_receivedBlock.contains(blockSeq)) {
            List<int> sendPacket = [
              [0, OpCode.ACK_VALUE],
              [data.data[2], data.data[3]],
            ].expand((x) => x).toList();
            _socket.send(
                sendPacket, InternetAddress(remoteAddress), remotePort);
            return;
          }
          _receivedBlock.add(blockSeq);
          if (_receivedBlock.length > 20) {
            _receivedBlock.removeAt(0);
          }

          List<int> d = data.data.sublist(4);
          if (d.isNotEmpty) {
            io.add(d);
          }
          List<int> sendPacket = [
            [0, OpCode.ACK_VALUE],
            [data.data[2], data.data[3]],
          ].expand((x) => x).toList();
          _socket.send(sendPacket, InternetAddress(remoteAddress), port);

          if (d.length < blockSize) {
            io.close();
            subscription.cancel();
            completer.complete(0);
          }
        }
        _checkError(data.data);
      }
    });

    List<int> sendPacket = [
      [0, OpCode.RRQ_VALUE],
      encoder.convert(remoteFile),
      [0],
      TransType.octet,
      [0],
    ].expand((x) => x).toList();
    _receivedBlock = <int>[];
    _socket.send(sendPacket, InternetAddress(remoteAddress), remotePort);

    return completer.future;
  }

  Future<int> _sendPacket(List<int> sendPacket, int blockSeq,
      String remoteAddress, int remotePort) async {
    Completer<int> completer = Completer();
    late StreamSubscription subscription;
    subscription = _stream.listen((ev) {
      if (RawSocketEvent.read == ev) {
        var data = _socket.receive();
        if (_checkAck(blockSeq, data!.data)) {
          subscription.cancel();
          completer.complete(data.port);
          return;
        }
        _checkError(data.data);
      }
    });
    _socket.send(sendPacket, InternetAddress(remoteAddress), remotePort);
    return completer.future;
  }

  bool _checkAck(int blockSeq, List<int> data) {
    if (data[0] << 8 | data[1] == OpCode.ACK_VALUE) {
      var ackValue = data[2] << 8 | data[3];
      if (blockSeq == ackValue) {
        return true;
      }
    }
    return false;
  }

  void _checkError(List<int> data) {
    if (data[0] << 8 | data[1] == OpCode.ERROR_VALUE) {
      var code = data[2] << 8 | data[3];
      List<int> msgData = [];
      for (int i = 4; i < data.length; ++i) {
        if (0 == data[i]) {
          break;
        }
        msgData.add(data[i]);
      }
      var msg = String.fromCharCodes(msgData);
      throw TFtpException(code, msg);
    }
  }

  void close() {
    _socket.close();
  }
}


example/lib/main.dart
import 'package:example/client.dart';
import 'package:example/server.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'TFtp Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'TFTP Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage>
    with SingleTickerProviderStateMixin {
  TabController controller;

  @override
  void initState() {
    super.initState();
    controller = new TabController(length: 2, vsync: this);
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        bottom: TabBar(
          controller: controller,
          tabs: [
            Tab(icon: Icon(Icons.wb_cloudy), child: Text("Server")),
            Tab(icon: Icon(Icons.phonelink_ring), child: Text("Client"))
          ],
        ),
      ),
      body: Center(
        child: TabBarView(
          controller: controller,
          children: [
            ServerDemo(),
            ClientDemo(),
          ],
        ),
      ),
    );
  }
}


lib/src/server.dart
import 'dart:async';
import 'dart:collection';
import 'dart:io';
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:tftp/tftp.dart';

typedef ReadFileCallBack = String Function(String file,
    void Function({ProgressCallback? progressCallback}) onProgress);
typedef WriteFileCallBack = String Function(
    String file,
    void Function(
            {bool overwrite,
            StreamTransformer<List<int>, dynamic>? transformer})
        doTransform);

class TFtpServer extends Stream<TFtpServerSocket> {
  TFtpServer(this.address, {this.port = 69}) {
    _controller = StreamController<TFtpServerSocket>(sync: true);
    RawDatagramSocket.bind(address, port).then(_init, onError: _initError);
  }

  static Future<TFtpServer> bind(address, {int port = 69}) async {
    return Future(() => TFtpServer(address, port: port));
  }

  late final HashMap<String, TFtpServerSocket> _socketDic = HashMap();
  late RawDatagramSocket _udpSocket;
  late StreamController<TFtpServerSocket> _controller;
  dynamic address;
  int port;

  void _init(RawDatagramSocket socket) {
    _udpSocket = socket;
    _udpSocket.listen((ev) {
      if (ev == RawSocketEvent.read) {
        var datagram = _udpSocket.receive();
        var tid = "${datagram?.address.toString()}:${datagram?.port}";
        if (null == _socketDic[tid]) {
          var socket =
              TFtpServerSocket(_udpSocket, datagram!.address, datagram.port);
          _socketDic[tid] = socket;
          _controller.add(socket);
        }
        _socketDic[tid]?.read(datagram!.data);
      }
    });
  }

  void _initError(error, stack) {
    close();
  }

  Future close({bool force = false}) {
    return _controller.close();
  }

  @override
  StreamSubscription<TFtpServerSocket> listen(
      void Function(TFtpServerSocket event)? onData,
      {Function? onError,
      void Function()? onDone,
      bool? cancelOnError}) {
    return _controller.stream.listen(onData,
        onError: onError, onDone: onDone, cancelOnError: cancelOnError);
  }
}

class TFtpServerSocket {
  TFtpServerSocket(this.socket, this.remoteAddress, this.remotePort,
      {this.blockSize = 512, this.onRead, this.onWrite, this.onError});

  RawDatagramSocket socket;
  InternetAddress remoteAddress;
  int remotePort;
  int blockSize;

  ReadFileCallBack? onRead;
  WriteFileCallBack? onWrite;
  ErrorCallBack? onError;

  late File? _writeFile;
  late IOSink? _writeSink;
  late StreamController<List<int>>? _writeStreamCtrl;
  late List<int>? _receivedBlock = <int>[];

  void read(List<int> data) {
    switch (data[0] << 8 | data[1]) {
      case OpCode.RRQ_VALUE:
        var info = _readFileNameAndTransType(data);
        if (kDebugMode) {
          print("read server file:${info.fileName} with ${info.transType}");
        }
        ProgressCallback? onReceive;
        if (null != onRead) {
          info.fileName = onRead!(info.fileName, ({progressCallback}) {
            onReceive = progressCallback;
          });
        }
        var file = File(info.fileName);
        if (!file.existsSync()) {
          throwError(Error.FILE_NOT_FOUND);
          return;
        }
        _write(file, onReceiveProgress: onReceive);
        return;
      case OpCode.WRQ_VALUE:
        var info = _readFileNameAndTransType(data);
        if (kDebugMode) {
          print("write server file:${info.fileName} with ${info.transType}");
        }

        if (null != _writeStreamCtrl) {
          _writeStreamCtrl?.close().then((_) {
            if (null != _writeSink) {
              _writeSink?.close();
              _writeSink = null;
            }
          });
        }
        if (null != _writeFile) {
          _writeFile?.deleteSync();
        }

        bool fileOverwrite = true;
        if (null != onWrite) {
          info.fileName =
              onWrite!(info.fileName, ({overwrite = true, transformer}) {
            fileOverwrite = overwrite;
            if (null != transformer) {
              _writeStreamCtrl?.stream.transform(transformer);
            }
          });
        }
        _writeFile = File(info.fileName);

        if (!fileOverwrite && _writeFile!.existsSync()) {
          throwError(Error.FILE_ALREADY_EXISTS);
          return;
        }
        try {
          _writeFile?.createSync();
        } catch (e) {
          throwError(Error.ACCESS_VIOLATION);
          return;
        }
        _writeStreamCtrl = StreamController(sync: true);
        _writeSink = _writeFile?.openWrite();
        _writeSink?.addStream(_writeStreamCtrl!.stream);

        _receivedBlock = <int>[];
        List<int> sendPacket = [
          [0, OpCode.ACK_VALUE],
          [0x00, 0x00],
        ].expand((x) => x).toList();
        socket.send(sendPacket, remoteAddress, remotePort);
        return;
      case OpCode.ACK_VALUE:
        throwError(Error.ILLEGAL_OPERATION);
        return;
      case OpCode.DATA_VALUE:
        if (null == _writeFile) {
          throwError(Error.ILLEGAL_OPERATION);
          return;
        }
        var blockSeq = (data[2] << 8) + data[3];
        if (_receivedBlock!.contains(blockSeq)) {
          List<int> sendPacket = [
            [0, OpCode.ACK_VALUE],
            [data[2], data[3]],
          ].expand((x) => x).toList();
          socket.send(sendPacket, remoteAddress, remotePort);
          return;
        }
        _receivedBlock!.add(blockSeq);
        if (_receivedBlock!.length > 20) {
          _receivedBlock?.removeAt(0);
        }
        var d = data.sublist(4);
        _writeStreamCtrl?.add(d);
        List<int> sendPacket = [
          [0, OpCode.ACK_VALUE],
          [data[2], data[3]],
        ].expand((x) => x).toList();
        socket.send(sendPacket, remoteAddress, remotePort);
        if (d.length < blockSize) {
          _writeStreamCtrl?.close().then((_) {
            _writeSink?.close();
            _writeSink = null;
          });
          _writeStreamCtrl = null;
          _writeFile = null;
        }
        return;
      case OpCode.ERROR_VALUE:
        var code = data[2] << 8 | data[3];
        List<int> msgData = [];
        for (int i = 4; i < data.length; ++i) {
          if (0 == data[i]) {
            break;
          }
          msgData.add(data[i]);
        }
        var msg = String.fromCharCodes(msgData);
        _getError(code, msg);
        return;
      default:
        throwError(Error.ILLEGAL_OPERATION);
        return;
    }
  }

  TransInfo _readFileNameAndTransType(List<int> data) {
    List<List<int>> packetData = List.generate(2, (index) => List.empty());

    int pos = -1;
    for (int val in data) {
      if (0 == val) {
        pos++;
      }
      if (pos > 1) {
        break;
      }
      packetData[pos].add(val);
    }

    List<int> fileNameData = packetData[0];
    List<int> transTypeData = packetData[1];
    return TransInfo(String.fromCharCodes(fileNameData, 2, fileNameData.length),
        String.fromCharCodes(transTypeData, 0, transTypeData.length));
  }

  /// 客户端读取文件时用于写入文件内容
  Future _write(File file, {ProgressCallback? onReceiveProgress}) async {
    RandomAccessFile fileWait2Write;
    int blockNum;
    Completer completer = Completer();

    RawDatagramSocket sendSocket = await RawDatagramSocket.bind(
        socket.address, Random().nextInt(10000) + 40000);
    fileWait2Write = await file.open();
    var totalSize = fileWait2Write.lengthSync();
    blockNum = 0;

    SendCompleter sendCompleter = SendCompleter();
    sendSocket.listen((ev) async {
      if (RawSocketEvent.read == ev) {
        var data = sendSocket.receive();
        if (data!.data[0] << 8 | data.data[1] == OpCode.ACK_VALUE) {
          var ackValue = data.data[2] << 8 | data.data[3];
          sendCompleter.completer!.complete(ackValue);
        }
      }
    });

    Future.microtask(() async {
      List<int> dataBlock;
      while ((dataBlock = await fileWait2Write.read(blockSize)).isNotEmpty) {
        blockNum++;
        blockNum = blockNum > 65535 ? 0 : blockNum;

        List<int> sendPacket = [
          [0, OpCode.DATA_VALUE],
          [blockNum >> 8, blockNum & 0xff],
          dataBlock
        ].expand((x) => x).toList();
        await _send(sendSocket, blockNum, sendPacket, onReceiveProgress,
            totalSize, sendCompleter);
        if (blockNum * blockSize == totalSize) {
          blockNum++;
          blockNum = blockNum > 65535 ? 0 : blockNum;
          List<int> sendPacket = [
            [0, OpCode.DATA_VALUE],
            [blockNum >> 8, blockNum & 0xff],
          ].expand((x) => x).toList();
          await _send(sendSocket, blockNum, sendPacket, onReceiveProgress,
              totalSize, sendCompleter);
        }
      }
      completer.complete();
    });

    return completer.future;
  }

  Future _send(
      RawDatagramSocket sendSocket,
      int blockNum,
      List<int> sendPacket,
      ProgressCallback? onReceiveProgress,
      int totalSize,
      SendCompleter sendCompleter) async {
    int ack;
    int sendTime = 0;
    do {
      sendTime++;
      sendSocket.send(sendPacket, remoteAddress, remotePort);
      if (null != onReceiveProgress) {
        var processed = blockNum * blockSize;
        processed = processed == totalSize ? processed - 1 : processed;
        processed = processed > totalSize ? totalSize : processed;
        onReceiveProgress(processed, totalSize);
      }
      sendCompleter.completer = Completer();
      ack = await sendCompleter.completer!.future.timeout(
        const Duration(seconds: 1),
      );
    } while (ack != blockNum && sendTime < 5);
  }

  void close() {}

  void throwError(int code) {
    if (null != onError) {
      onError!(code, errorDic[code]!);
    }
    List<int> sendPacket = [
      [0, OpCode.ERROR_VALUE],
      [code >> 8, code & 0xff],
      encoder.convert(errorDic[code]!),
    ].expand((x) => x).toList();

    socket.send(sendPacket, remoteAddress, remotePort);
  }

  void _getError(int code, String message) {
    if (null != onError && code != 0) {
      onError!(code, message);
    }
  }

  void listen({
    ReadFileCallBack? onRead,
    WriteFileCallBack? onWrite,
    ErrorCallBack? onError,
  }) {
    this.onRead = onRead;
    this.onWrite = onWrite;
    this.onError = onError;
  }
}

class SendCompleter {
  Completer<int>? completer;
}

enum WorkType { IDLE, READ, WRITE }
